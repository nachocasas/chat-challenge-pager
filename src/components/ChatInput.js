import React, { Component } from 'react';
export default class ChatInput extends Component {

  constructor(props){
    super(props);
    this.state = {
      text: '',
    }
    this.tipingTimeOut = null;
    this.handleChange = this.handleChange.bind(this);
    this.handleEnter = this.handleEnter.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {
    let self = this;
    this.setState({
      text: e.target.value
    })

    this.props.isTyping(true)
    if(this.tipingTimeOut){ 
      clearTimeout(this.tipingTimeOut);
    }

    this.tipingTimeOut = setTimeout(function() { self.props.isTyping(false) }, 2000);
  }

  handleEnter(e){
    const code = (e.keyCode ? e.keyCode : e.which)
    if(code == 13){
      this.handleSubmit();
    }
  }

  handleSubmit(e){
    if(this.state.text.trim() != ''){
      this.props.onEnter(this.state.text);
      this.setState({ text: '' });
    }
  }

  render(){
    return(
      <div className="chatinput">
        <input onChange={ this.handleChange } onKeyPress={ this.handleEnter } value={this.state.text} />
        <button onClick={ this.handleSubmit }>Send</button>
      </div>
    )
  }
}