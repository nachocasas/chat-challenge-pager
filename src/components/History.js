import React, { Component } from 'react';
import { connect } from 'react-redux';

class History extends Component {
  
  constructor(props){
    super(props);
  }
 
  componentDidUpdate(){
    this.divContainer.scrollTop = this.divContainer.scrollHeight - this.divContainer.clientHeight;
  }

  renderChatHistory(history){
    return history.map((item, i) => {
      return <li key={ i }><span>{item.username}</span>: {item.data} </li>
    })
  }

  render(){
    const chatHistory = this.renderChatHistory(this.props.history);
    
    return(
      <div ref={ (ref) => this.divContainer = ref } className="history">
        <ul>
          { chatHistory }
        </ul>
      </div>
    )
  }
}

function mapStateToProps({ chat }){
  if(!chat.history) {
      return { history: [] };
  }
  return { history: chat.history };
}

export default connect(mapStateToProps)(History);