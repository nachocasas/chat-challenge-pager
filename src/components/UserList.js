import React, { Component } from 'react';
import { connect } from 'react-redux';

class UserList extends Component {
  
 
  renderChatHistory(userlist){
    let styleClass = '';
    return userlist.map((item, i) => {
      styleClass = i%2 ? 'even' : ''
      styleClass = item==this.props.username ? styleClass + ' current-user' : styleClass;
      return <li className={styleClass} key={ i }>{item}</li>
    })
  }
  
  render(){
    const userList = this.renderChatHistory(this.props.userlist);

    return(
      <div className="userlist">
        <h3>userlist</h3>
        <ul>{ userList }</ul>
      </div>
    )
  }
}

function mapStateToProps({ chat }){
  if(!chat.userlist) {
      return { userlist: [] };
  }
  return { userlist: chat.userlist, username: chat.username };
}

export default connect(mapStateToProps)(UserList);