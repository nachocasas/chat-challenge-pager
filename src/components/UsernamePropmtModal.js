import React, { Component } from 'react';

export default class UsernamePropmtModal extends Component {

  constructor(props){
    super(props);
    this.state = {
      text: ''
    }

    this.handleChange = this.handleChange.bind(this);
    this.handleEnter = this.handleEnter.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e) {
    this.setState({
      text: e.target.value
    })
  }

  handleEnter(e){
    const code = (e.keyCode ? e.keyCode : e.which)
    if(code == 13){
      this.handleSubmit();
    }
  }

  handleSubmit(){
    this.props.onUsernameEnter(this.state.text);
    this.setState({ text: '' });
  }

  render(){
    return(
      <div className="username-modal">
        <div className='prompt'>
          <h3>Please enter your username:</h3>
          <input onChange={ this.handleChange } onKeyPress={ this.handleEnter }></input>
          <button onClick={ this.handleSubmit }>Chat!</button>
        </div>
      </div>
    )
  }
}