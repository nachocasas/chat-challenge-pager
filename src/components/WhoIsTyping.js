import React from 'react';

export function WhoIsTyping (props){
  let typing = '';
  if(props.names != '') {
    typing = `${props.names} is typing...`;
  }
  
  return (
    <div className="whoistyping">
      {typing}
    </div>
  )
  
}