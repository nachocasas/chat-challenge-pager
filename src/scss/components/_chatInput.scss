.chatinput {
  width: 100%;
  padding: 1% 0;
  
  input {
    width: 100%;
    padding:1%;
    border:0;
    border-radius: 10px;
    outline: 0;
  }

}