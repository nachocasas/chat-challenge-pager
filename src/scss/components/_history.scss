.history {
  width: 100%;
  padding: 1%;
  height: 70vh;
  overflow: hidden;
  background-color: $history-background-color;
  font-size: 0.9rem;

  &:hover {
    overflow-y: auto;
  }

  ul {
    list-style-type: none;
    margin:0;
    padding: 0;
  }

  li {
    margin: 0.5% 0;
    color: $history-text-color;
  }

  span {
    font-weight: bold;
    color: $history-username-color;
  }
}