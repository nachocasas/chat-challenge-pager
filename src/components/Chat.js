import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setUsername,
          handleConnectionStatusEvents, 
          setIncommingHistoryEvent,
          setIncommingUserlistEvent,
          setIncommingMessageEvent,
          setIncommingIsTypingEvent } from '../actions'
import io from "socket.io-client"

import History from './History';
import UserList from './UserList';
import ChatInput from './ChatInput';
import UsernamePropmtModal from './UsernamePropmtModal';
import { WhoIsTyping } from './WhoIsTyping'

const SOCKET_URL = 'http://localhost:3000'

class Chat extends Component {
  
  constructor(props){
    super(props);
    this.socket = null;
    this.setUserAndConnect = this.setUserAndConnect.bind(this);
    this.sendMessage = this.sendMessage.bind(this);
    this.setTyping = this.setTyping.bind(this);
  }

  setUserAndConnect(username){
    this.socket = io.connect(SOCKET_URL, { query: 'username='+username});
    this.props.setUsername(username);
    this.fireEventListeners();
  }

  sendMessage(message){
    this.socket.emit('message', message );
  }

  setTyping(isTyping){
    this.socket.emit('isTyping', isTyping );
  }

  whoIsTyping(){
    return this.props.isTyping ? this.props.isTyping.join(', ') : null
  }

  fireEventListeners(){
    const { props, socket } = this;
    props.handleConnectionStatusEvents(socket);
    props.setIncommingHistoryEvent(socket);
    props.setIncommingUserlistEvent(socket);
    props.setIncommingMessageEvent(socket);
    props.setIncommingIsTypingEvent(socket);
  }

  render(){
    const whoIsTyping = this.whoIsTyping();
    return(
      <div className="chat-app">
        { !this.props.connected && <UsernamePropmtModal onUsernameEnter = { this.setUserAndConnect } /> }
        <div className="header">
          <h1>Pager challenge chat</h1>
        </div>
        <div className="main-content">
          <div className="left-container">
            <History />
            <WhoIsTyping names={ whoIsTyping } />
            <ChatInput isTyping={ this.setTyping } onEnter={ this.sendMessage } />
          </div>
          <div className="right-container">
            <UserList onUserClick={ this.handleUserSelect } />
          </div>
        </div>
      </div>
    )
  }
}

function mapStateToProps({ chat }) {
  const { username, connected, isTyping } = chat;
  return { username, connected, isTyping }
}


export default connect(mapStateToProps, {
  setUsername,
  handleConnectionStatusEvents, 
  setIncommingHistoryEvent,
  setIncommingUserlistEvent,
  setIncommingMessageEvent, 
  setIncommingIsTypingEvent })(Chat)