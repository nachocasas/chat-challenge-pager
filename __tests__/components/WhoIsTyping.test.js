import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { WhoIsTyping } from '../../src/components/WhoIsTyping';

describe('<WhoIsTyping />', () => {
  describe('render()', () => {

    test('render component', () => {
      const wrapper = shallow(<WhoIsTyping />);

      expect(toJson(wrapper)).toMatchSnapshot();
    });

  });
});