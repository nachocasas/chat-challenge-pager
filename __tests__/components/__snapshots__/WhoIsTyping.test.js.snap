// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WhoIsTyping /> render() render component 1`] = `
<div
  className="whoistyping"
>
  undefined is typing...
</div>
`;
